disk/test/com/treode/disk/edit/DiskIOSpec.scala

```diff
@@ -18,62 +18,150 @@ package com.treode.disk.edit
+    val posB = dsp.write(pDesc, 0, 0, b ) .expectPass()
+    val readA = dr.read(pDesc, posA).expectPass()
+    val readB = dr.read(pDesc, posB).expectPass()
     assert (a.equals (readA))
     assert (b.equals (readB))
-    assert (posA == startPos + lenA)
-    assert (posB == posA + lenB)
+    assert (posA.offset == 0 )
+    assert (posB.offset == a.length + 1)
   }
 
   it should "write and read correctly out of order" in {
     implicit val scheduler = StubScheduler.random()
+
+    val stringPickler = 
+        wrap(string)
+        .build(x => x )
+        .inspect( x=> x )
+
     val a = "abcdef"
     val b = "123456789"
     val startPos = 0
     val f = StubFile (1 << 20, 0)
-
-    val dw = new PageWriter (f)
+    val dsp = new PageDispatcher
+    val dw = new PageWriter (dsp,f)
     val dr = new PageReader (f)
-    val (posA, lenA) = dw.write (a) .expectPass()
-    val (posB, lenB) = dw.write (b) .expectPass()
-    val readB = dr.readString (posA,     lenB) .expectPass()
-    val readA = dr.readString (startPos, lenA) .expectPass()
+    val pDesc = PageDescriptor(0x25, stringPickler)
+    val posB = dsp.write(pDesc, 0, 0, b ) .expectPass()
+    val posA = dsp.write(pDesc, 0, 0, a ) .expectPass()
+    val readA = dr.read(pDesc, posA).expectPass()
+    val readB = dr.read(pDesc, posB).expectPass()
     assert (a.equals (readA))
     assert (b.equals (readB))
-    assert (posA == startPos + lenA)
-    assert (posB == posA + lenB)
+    assert (posA.offset == b.length + 1)
+    assert (posB.offset == 0)
+  }
+
+  it should "be able to write a full batch out of order correctly and then read it " in {
+    implicit val scheduler = StubScheduler.random()
+    val stringPickler = 
+        wrap(string, string)
+        .build(x => (x._1, x._2) )
+        .inspect( x=> (x._1, x._2) )
+    val f = StubFile (1 << 20, 0)
+    val a = "lorem "
+    val b = "ipsum "
+    val c = "dolor "
+    val d = "sit "
+    val dsp = new PageDispatcher
+    val pDesc = PageDescriptor(0x25, stringPickler)
+    //the following are purposely done out of order
+    val ad_callback = dsp.write(pDesc, 0, 0, (a,d) ).capture()
+    val bc_callback = dsp.write(pDesc, 0, 0, (b,c) ).capture()
+    val dw = new PageWriter (dsp, f)
+    val posAD = ad_callback.expectPass()
+    val posBC = bc_callback.expectPass()
+    val dr = new PageReader (f)
+    val readBC = dr.read(pDesc, posBC).expectPass()
+    val readAD = dr.read(pDesc, posAD).expectPass()
+    assert (a.equals(readAD._1))
+    assert (b.equals(readBC._1))
+    assert (c.equals(readBC._2))
+    assert (d.equals(readAD._2))
+    
+    assert ( (posAD.disk, posAD.offset, posAD.length)  == (0, 0, 12) )
+    assert ( (posBC.disk, posBC.offset, posBC.length)  == (0, 12, 14) ) 
+
+  }
+
+
+  it should "be able to write mixed types correctly and then read it " in {
+    implicit val scheduler = StubScheduler.random()
+    val multiPickler = 
+        wrap(string, long, int)
```

Comment:
Testing with just the pickler `string` and `int` is sufficient. Testing with complex picklers is not testing the page mechanism's handling of picklers, it is testing the picklers, which are already tested in their own specs, .